Add ability to build images within Openstack (#1013)
kubernetes-sigs/image-builder: docs/book/src/capi/providers/openstack-remote.md

# Building Images on OpenStack

## Hypervisor

> The image is built using Openstack.

Reviewer: ```suggestion
The image is built using OpenStack.
```

### Prerequisites for Openstack builds

Check any prerequisites over on the [Packer docs for the Openstack builder](https://developer.hashicorp.com/packer/plugins/builders/openstack).

> Also ensure you have a [Ubuntu 20.04](https://cloud-images.ubuntu.com/focal/current/) or [Ubuntu 22.04](https://cloud-images.ubuntu.com/jammy/current/) cloud image available in your Openstack instance before continuing as it will need to be referenced.

Reviewer: ```suggestion
Also ensure that you have a [Ubuntu 20.04](https://cloud-images.ubuntu.com/focal/current/) or [Ubuntu 22.04](https://cloud-images.ubuntu.com/jammy/current/) cloud image available in your OpenStack instance before continuing as it will need to be referenced.
```

> This also supports Flatcar - so far only Stable has been tested.

Reviewer: ```suggestion
This build process also supports Flatcar Linux, but only Stable has been tested.
```

#### Note

> > Other OS's could be supported and additions are welcome.

Reviewer: ```suggestion
> Other operating systems could be supported and additions are welcome.
```

## Setup Openstack authentication

> Ensure you have set up your method of authentication - [examples here](https://docs.openstack.org/python-openstackclient/zed/cli/authentication.html).

Reviewer: ```suggestion
Ensure you have set up your method of authentication. See the [examples here](https://docs.openstack.org/python-openstackclient/zed/cli/authentication.html).
```

You can also check out the [packer builder](https://developer.hashicorp.com/packer/plugins/builders/openstack#configuration-reference) for more information on authentication.

> You should be able to run commands against openstack before running this builder otherwise it will fail.

Reviewer: ```suggestion
You should be able to run commands against OpenStack before running this builder, otherwise it will fail.
```

You can test with a simple command such as `openstack image list`.

## Building Images

The build [prerequisites](../capi.md#prerequisites) for using `image-builder` for

> building Openstack images are managed by running:

Reviewer: ```suggestion
building OpenStack images are managed by running:
```

```bash
cd image-builder/images/capi
make deps-openstack
```

> ### Define variables for Openstack build

Reviewer: ```suggestion
### Define variables for OpenStack build
```

Using the [Openstack packer provider](https://developer.hashicorp.com/packer/plugins/builders/openstack), an instance will be deployed and an image built from it.

> A certain set of environment variables (example below) must be defined in a json file and referenced as shown below in the build example.

Reviewer: ```suggestion
A certain set of environment variables must be defined in a json file and referenced as shown below in the build example.
```

Replace UPPERCASE variables with your values.
```json
{
  "source_image": "OPENSTACK_SOURCE_IMAGE_ID",
  "networks": "OPENSTACK_NETWORK_ID",
  "flavor": "OPENSTACK_INSTANCE_FLAVOR_NAME",
  "floating_ip_network": "OPENSTACK_FLOATING_IP_NETWORK_NAME",
  "image_name": "KUBE-UBUNTU",
  "image_visibility": "public",
  "image_disk_format": "raw",
  "volume_type": "",
  "ssh_username": "ubuntu"
}
```

Check out `images/capi/packer/openstack/packer.json` for more variables such as allowing the use of floating IPs and config drives.

> ### Building Image on Openstack

Reviewer: ```suggestion
### Building Image on OpenStack
```

From the `images/capi` directory, run `PACKER_VAR_FILES=var_file.json make build-openstack-<DISTRO>`.

An instance is built in Openstack from the source image defined and once completed, the instance is shutdown and the image gets created.
This image will default to private however this can be controlled via `image_visibility`.

> For building a ubuntu-2204 based capi image with Kubernetes 1.23.15, run the following commands:

Reviewer: ```suggestion
For building a ubuntu 22.04-based CAPI image with Kubernetes 1.23.15, run the following commands:
```

# Building Images on OpenStack

## Hypervisor

The image is built using Openstack.

### Prerequisites for Openstack builds

Execute the following command to install the Openstack client.
```bash
# pip install python-openstackclient
```
Alternatively, you can [download and install the tarball](https://docs.openstack.org/python-openstackclient/latest/#getting-started).

> 

Reviewer 2: are you sure that the Packer client needs the Python OpenStack client to work?

Author: That's a good point actually. It's only really recommended by the builder docs for troubleshooting auth issues so I can get that removed. I'll try and get your points actioned today or tomorrow.

Also ensure you have a [Ubuntu 20.04](https://cloud-images.ubuntu.com/focal/current/) or [Ubuntu 22.04](https://cloud-images.ubuntu.com/jammy/current/) cloud image available in your Openstack instance before continuing as it will need to be referenced.

#### Note

> > Other OS's will be supported at a later time.

Reviewer 2: our team (cc @okozachenko1203) would want to get Flatcar support in out of the box, we will work on a PR to your branch.

Author: Sounds good. We have no initial requirement right now but it would be good to see other distros supported.

> ## Setup Openstack authentication

Reviewer 2: Would it make sense to instead link to https://developer.hashicorp.com/packer/plugins/builders/openstack#configuration-reference and mention that since we're using Packer?

The build [prerequisites](../capi.md#prerequisites) for using `image-builder` for
building Openstack images are managed by running:

```bash
cd image-builder/images/capi
make deps-openstack
```

### Define variables for Openstack build
Using the [Openstack packer provider](https://developer.hashicorp.com/packer/plugins/builders/openstack), an instance will be deployed and an image built from it.
A certain set of environment variables (example below) must be defined in a josn file  and reference it during `make build-openstack-ubuntu-xxxx`. Please replace xxxx with 2004 or 2204.

Replace UPPERCASE variables with your values.
```json
{
  "source_image": "OPENSTACK_SOURCE_IMAGE_ID",
  "networks": "OPENSTACK_NETWORK_ID",
  "flavor": "OPENSTACK_INSTANCE_FLAVOR_NAME",
  "floating_ip_network": "OPENSTACK_FLOATING_IP_NETWORK_NAME"
}
```

#### Note:
> The following Kuberentes versions have been tested: 1.23.10, 1.24.7 and 1.25.3. <br>
The following crictl versions have been tested: 1.23.0, 1.24.0 and 1.25.0.

Check out `images/capi/packer/openstack/packer.json` for more variables such as allowing the use of floating IPs and config drives.

### Building Image on Openstack

> From the `images/capi` directory, run `PACKER_VAR_FILES=var_file.json make build-openstack-ubuntu-xxxx`.

Reviewer 2: I would un-hardcode the ubuntu bits here since we will have (hopefully) Flatcar.

Author: Yeah no problem. Initially it was there because Ubuntu was the only one supported. When you get Flatcar in, if you want to change that too that would be great. For now I left it as Ubuntu as it's the only supported distro on this PR.

> An instance is built in Openstack from the source image defined and once completed, the instance is shutdown and the volume uploaded into an image that can then be used.

Reviewer 2: ```suggestion
An instance is built in Openstack from the source image defined and once completed, the instance is shutdown and the image gets created that can then be used afterwards.
```

Author: I'm happy with that change of wording - I'll take it in.

> This image will default to private and will need to be set as shared or public to be used by other projects within Openstack.

Reviewer 2: You should mention that `image_visibility` can control this.
